### PR TITLE
[READY FOR REVIEW] Lava land generation overhaul / variety

### PR DESCRIPTION
**Pull request by Qwertytoforty**
## What Does This PR Do

lavaland now generates with either lava, liquid plasma, or chasms. Mapping helper added to this.
Lava rivers change generation depending on "lava" type.
Chasms can now be built on properly
Bridges can generate across rivers.

## Why It's Good For The Game

See the post on discord, but also:
I am planning to add some variety to lavaland, in some various forms.

The goal of this project is to add more variety to lavalands generation, and open up branches for future expansion.

Currently this PR is 3 goals.

Generate lava in 3 forms. Lavaland lavaland. A liquid plasma lavaland. A chasm rocky lavaland. (todo: resprite chasms to have a bit of a glow to be more noticable)

Adjust lava river generation. Plasma rivers will be diffrently shaped than lava rivers, and chasms will be much thiner, and occasionally generate bridges across them. 

Adjust cave generation, adding some new types, such as thinner caves packed with more ash inbetween sections, wider caves, caves that double back more, or caves that lead into larger open rooms, that will have more foes in the open area. 1 cave type will be prioritied in a lavaland generation, with some caves occasionally being a diffrent style.

## TODO:

- [x] Easier to see chasms, upgrade to mesons to make chasms highlighted?
- [x] Bridges for chasm generation / labour camp
- [x] Adjust cave generation for new cave types
- [x] defines instead of "name here"

## Images of changes

See discord post where I can upload big boy files

## Testing

I generated lavaland a bunch.

## Changelog
:cl:
tweak: Lavaland can now generate with liquid plasma instead of lava, or chasms. Watch your step!
tweak: The rivers in lavaland will vary in generation rules depending on lava / plasma / chasm.
add: Bridges may generate across the rivers.
tweak: 2 new generation rulesets added for cave generation. Be on the lookout for the Blocked Burrows, more crouded caves filled with rocks and plants, and Deadly Deeprock, with more open cavern rooms filled with monsters.
fix: You can build across chasms properly now
/:cl:

### CODE REVIEW

**Comment by tipoima** [comment]
Oh boy, can't wait to fall in chasms constantly.

How will liquid plasma be different from lava aside from the river shape?

**Comment by Qwertytoforty** [comment]
> Oh boy, can't wait to fall in chasms constantly.
> 
> How will liquid plasma be different from lava aside from the river shape?

Apart from river shape, one can scoop it up for liquid plasma in a beaker, it does less burn damage but more toxin damage, making it less punishing to plasmamen and IPCs, but more punishing to other species.

**Comment by TravisAngeI** [comment]
if im reading the pr right it replaces all the lava pools with different types of 'ouchie ow kill stuff', wouldn't that mean that the gulag camp would be unescapable because you'll just fall in a chasm?

**Comment by tipoima** [comment]
> if im reading the pr right it replaces all the lava pools with different types of 'ouchie ow kill stuff', wouldn't that mean that the gulag camp would be unescapable because you'll just fall in a chasm?

"Finally. Containable clings."

**Comment by Qwertytoforty** [comment]
> 

Chasms can be built over, by applying metal rods to the chasm, then a floor tile.

This makes the chasms actually *more* escape able than lava. In cases like that, security would have to adapt and not send perma prisoners there, use it for lesser crimes.

**Comment by S34NW** [comment]
I think your meson changes should be a separate PR honestly

File: code/controllers/subsystem/non_firing/SSmapping.dm [hunk 1]
```diff
@@ -63,8 +71,22 @@ SUBSYSTEM_DEF(mapping)
...
+			if(/turf/simulated/floor/plating/lava/smooth/lava_land_surface/plasma) //More rivers, smaller
+				spawn_rivers(level_name_to_num(MINING), nodes = 2)
+				spawn_rivers(level_name_to_num(MINING), nodes = 2)
+			if(/turf/simulated/floor/chasm/straight_down/lava_land_surface) //Thiner chasms, bridges, reaches to edge of map.
+				spawn_rivers(level_name_to_num(MINING), nodes = 6, turf_type = /turf/simulated/floor/plating/lava/smooth/mapping_lava, whitelist_area = /area/lavaland/surface/outdoors, min_x = 50, min_y = 7, max_x = 250, max_y = 225, prob = 10, prob_loss = 5)
+		var/time_spent = stop_watch(lavaland_setup_timer)
+		log_startup_progress("Successfully populated lavaland in [time_spent]s.")
+		if(time_spent >= 10)
+			log_startup_progress("!!!ERROR!!!! Lavaland took FAR too long to generate at [time_spent] seconds. Notify maintainers imediately! !!!ERROR!!!") //In 3 testing cases so far, I have had it take far too long to generate. I am 99% sure I have fixed this issue, but never hurts to be sure
+			WARNING("!!!ERROR!!!! Lavaland took FAR too long to generate at [time_spent] seconds. Notify maintainers imediately! !!!ERROR!!!")
```
Comment (Henri215) [inline code/controllers/subsystem/non_firing/SSmapping.dm:86]:
```suggestion
			log_startup_progress("!!!ERROR!!! Lavaland took FAR too long to generate at [time_spent] seconds. Notify maintainers immediately! !!!ERROR!!!") //In 3 testing cases so far, I have had it take far too long to generate. I am 99% sure I have fixed this issue, but never hurts to be sure
			WARNING("!!!ERROR!!! Lavaland took FAR too long to generate at [time_spent] seconds. Notify maintainers immediately! !!!ERROR!!!")
```

File: code/datums/ruins/bridges/bridges.dm [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	. = ..()
+
+	var/turf/east = locate(x+3,y,z)
+	var/turf/west = locate(x-3,y,z)
+	var/turf/north = locate(x,y+3,z)
+	var/turf/south = locate(x,y-3,z)
+	var/turf/e1 = locate(x+1,y,z)
+	var/turf/w1 = locate(x-1,y,z)
+	var/turf/n1 = locate(x,y+1,z)
+	var/turf/s1 = locate(x,y-1,z)
```
Comment (Henri215) [inline code/datums/ruins/bridges/bridges.dm:73]:
```suggestion
	var/turf/east = locate(x + 3, y, z)
	var/turf/west = locate(x - 3, y, z)
	var/turf/north = locate(x, y + 3, z)
	var/turf/south = locate(x , y - 3, z)
	var/turf/e1 = locate(x + 1, y, z)
	var/turf/w1 = locate(x - 1, y, z)
	var/turf/n1 = locate(x, y + 1, z)
	var/turf/s1 = locate(x, y - 1, z)
```

File: code/game/turfs/simulated/floor/asteroid_floors.dm [hunk 3]
```diff
@@ -198,7 +198,9 @@ GLOBAL_LIST_INIT(megafauna_spawn_list, list(/mob/living/simple_animal/hostile/me
 	if (!megafauna_spawn_list)
 		megafauna_spawn_list = GLOB.megafauna_spawn_list
 	if (!flora_spawn_list)
-		flora_spawn_list = list(/obj/structure/flora/ash/leaf_shroom = 2 , /obj/structure/flora/ash/cap_shroom = 2 , /obj/structure/flora/ash/stem_shroom = 2 , /obj/structure/flora/ash/cacti = 1, /obj/structure/flora/ash/tall_shroom = 2)
+		flora_spawn_list = list(/obj/structure/flora/ash/leaf_shroom = 2 , /obj/structure/flora/ash/cap_shroom = 2 , /obj/structure/flora/ash/stem_shroom = 2 , /obj/structure/flora/ash/cacti = 1, /obj/structure/flora/ash/tall_shroom = 2, /obj/structure/flora/ash/rock/style_random = 1)
```
Comment (Henri215) [inline code/game/turfs/simulated/floor/asteroid_floors.dm:201]:
```suggestion
		flora_spawn_list = list(/obj/structure/flora/ash/leaf_shroom = 2, /obj/structure/flora/ash/cap_shroom = 2, /obj/structure/flora/ash/stem_shroom = 2, /obj/structure/flora/ash/cacti = 1, /obj/structure/flora/ash/tall_shroom = 2, /obj/structure/flora/ash/rock/style_random = 1)
```

File: code/game/turfs/simulated/floor/lava.dm [hunk 4]
```diff
@@ -142,3 +142,100 @@
...
+	for(var/thing in thing_to_check)
+		if(isobj(thing))
+			var/obj/O = thing
+			if(!O.simulated)
+				continue
+			if((O.resistance_flags & (LAVA_PROOF|INDESTRUCTIBLE)) || O.throwing)
+				continue
+			. = 1
+			if((O.resistance_flags & (ON_FIRE)))
+				continue
```
Comment (Henri215) [inline code/game/turfs/simulated/floor/lava.dm:187]:
```suggestion
			if((O.resistance_flags & ON_FIRE))
				continue
```

File: code/game/turfs/simulated/floor/lava.dm [hunk 5]
```diff
@@ -142,3 +142,100 @@
...
+				if("lava" in live.weather_immunities)
+					continue
+			if("lava" in burn_living.weather_immunities)
+				continue
+			burn_living.adjustFireLoss(2)
+			if(QDELETED(burn_living))
+				return
+			burn_living.adjust_fire_stacks(20) //dipping into a stream of plasma would probably make you more flammable than usual
+			burn_living.IgniteMob()
+			burn_living.adjust_bodytemperature(-rand(50,65)) //its cold, man
```
Comment (Henri215) [inline code/game/turfs/simulated/floor/lava.dm:219]:
```suggestion
			burn_living.adjust_bodytemperature(-rand(50, 65)) //its cold, man
```

File: code/modules/mining/lavaland/ash_flora.dm [hunk 6]
```diff
@@ -143,6 +147,40 @@
...
+
+/*********
+ * Rocks *
+ *********/
+// (I know these aren't plants)
+
+/obj/structure/flora/ash/rock
+	name = "large rock"
+	icon_state = "basalt1"
+	desc = "A volcanic rock. Pioneers used to ride these babies for miles."
```
Comment (Henri215) [inline code/modules/mining/lavaland/ash_flora.dm:159]:
```suggestion
/obj/structure/flora/ash/rock
	name = "large rock"
	desc = "A volcanic rock. Pioneers used to ride these babies for miles."
	icon_state = "basalt1"
```

File: code/modules/mining/lavaland/ash_flora.dm [hunk 7]
```diff
@@ -143,6 +147,40 @@
...
+	density = TRUE
+	resistance_flags = FIRE_PROOF
+	harvest = /obj/item/stack/ore/glass/basalt
+	harvest_time = 6 SECONDS
+	harvest_amount_low = 10
+	harvest_amount_high = 20
+	harvest_message_low = "You finish mining the rock."
+	harvest_message_med = "You finish mining the rock."
+	harvest_message_high = "You finish mining the rock."
```
Comment (Henri215) [inline code/modules/mining/lavaland/ash_flora.dm:168]: Going to keep these all the same?

File: code/game/turfs/simulated/floor/lava.dm [hunk 8]
```diff
@@ -142,3 +142,100 @@
...
+	if(AM)
+		thing_to_check = list(AM)
+	for(var/thing in thing_to_check)
+		if(isobj(thing))
+			var/obj/O = thing
+			if(!O.simulated)
+				continue
+			if((O.resistance_flags & (LAVA_PROOF|INDESTRUCTIBLE)) || O.throwing)
+				continue
+			. = 1
```
Comment (S34NW) [inline code/game/turfs/simulated/floor/lava.dm:185]:
```suggestion
			. = TRUE
```
A few instances of this, please fix

File: code/game/turfs/simulated/floor/lava.dm [hunk 9]
```diff
@@ -142,3 +142,100 @@
...
+				continue
+			if(!(O.resistance_flags & FLAMMABLE))
+				O.resistance_flags |= FLAMMABLE //Even fireproof things burn up in lava
+			if(O.resistance_flags & FIRE_PROOF)
+				O.resistance_flags &= ~FIRE_PROOF
+			if(O.armor.getRating(FIRE) > 50) //obj with 100% fire armor still get slowly burned away.
+				O.armor = O.armor.setRating(fire_value = 50)
+			O.fire_act(10000, 1000)
+
+		if(isliving(thing))
```
Comment (S34NW) [inline code/game/turfs/simulated/floor/lava.dm:196]:
```suggestion
		if(isliving(thing))
```
this last if in the for loop can be turned into a guard clause with a continue

File: code/game/turfs/simulated/floor/asteroid_floors.dm [hunk 10]
```diff
@@ -314,7 +375,11 @@ GLOBAL_LIST_INIT(megafauna_spawn_list, list(/mob/living/simple_animal/hostile/me
 #undef SPAWN_BUBBLEGUM
 
 /turf/simulated/floor/plating/asteroid/airless/cave/proc/SpawnFlora(turf/T)
-	if(prob(12))
+	var/floraprob = 12
+	switch(SSmapping.cave_theme)
+		if(BLOCKED_BURROWS)
+			floraprob = 30 //Lots of folliage, lots of blockage
```
Comment (S34NW) [inline code/game/turfs/simulated/floor/asteroid_floors.dm:381]: switch with one thing in it 😠

File: code/game/turfs/simulated/floor/asteroid_floors.dm [hunk 11]
```diff
@@ -291,11 +346,17 @@ GLOBAL_LIST_INIT(megafauna_spawn_list, list(/mob/living/simple_animal/hostile/me
...
 				randumb = pickweight(mob_spawn_list)
-
-		for(var/thing in urange(12, T)) //prevents mob clumps
+		var/scanrange = 12
+		var/megafaunarange = 7
+		switch(SSmapping.cave_theme)
+			if(DEADLY_DEEPROCK)
+				if(prob(50) && monsterprob > 30)
+					scanrange = rand(4, 7)
+					megafaunarange = scanrange
```
Comment (S34NW) [inline code/game/turfs/simulated/floor/asteroid_floors.dm:355]: switch with one thing in it 😠

File: code/game/turfs/simulated/floor/asteroid_floors.dm (hunk 10 above)
Comment (Qwertytoforty) [reply code/game/turfs/simulated/floor/asteroid_floors.dm:381]: Do we want to make this switches so it is modular and easy to change, or do we want to remove switches so further expansions to generation have to add them back?

File: code/game/turfs/simulated/floor/lava.dm (hunk 8 above)
Comment (Qwertytoforty) [reply code/game/turfs/simulated/floor/lava.dm]: Done!

File: code/game/turfs/simulated/floor/lava.dm (hunk 9 above)
Comment (Qwertytoforty) [reply code/game/turfs/simulated/floor/lava.dm]: Done!

**Review by S34NW — APPROVED** [review]
time to see what this breaks

**Pull request merged**